File: 卡码/代码随想录/1/01-数组/C++/05-螺旋矩阵/LeetCode_0059.cpp
//59. ÂİĞı¾ØÕó II
#include <iostream>
#include <vector>
using namespace std;

class Solution {
	public:
		vector<vector<int>> generateMatrix(int n) {
			int startX=0,startY=0,offset=1,count=1,loop=n/2;
			vector<vector<int>> nums(n);
			for(auto it=nums.begin();it!=nums.end();it++){
				vector<int> tmp(n);
				(*it)=tmp;
			}
			while(loop--){
				int i=startX,j=startY;
				for(j=startY;j<n-offset;j++) nums[i][j]=count++;
				for(i=startX;i<n-offset;i++) nums[i][j]=count++;
				for(;j>startY;j--) nums[i][j]=count++;
				for(;i>startX;i--) nums[i][j]=count++;
				startX++,startY++,offset++;
			}
			if(n%2!=0) nums[n/2][n/2]=count;
			return nums;
		}
		void printNums(vector<vector<int>> nums){
			for(int i=0;i<nums.size();i++){
				for(int j=0;j<nums.size();j++){
					if(j!=0) printf(" ");
					printf("%d",nums[i][j]);
				}
				printf("\n");
			}
		}
};

int main() {
	Solution* solution=new(Solution);
	vector<vector<int>> nums=solution->generateMatrix(5);
	solution->printNums(nums);
	return 0;
}

File: 卡码/代码随想录/1/01-数组/Python/05-螺旋矩阵/LeetCode__0059.py
#59. 螺旋矩阵 II
from typing import *

class Solution:
	def generateMatrix(self, n: int) -> List[List[int]]:
		startX:int=0
		startY:int=0
		count:int=1
		offset:int=1
		loop:int=n//2
		nums:List[List[int]]=[[0 for x in range(n)] for x in range(n)]
		while loop>0:
			i:int=startX
			j:int=startY
			for j in range(startY,n-offset):
				nums[i][j]=count
				count+=1
			j+=1
			for i in range(startX,n-offset):
				nums[i][j]=count
				count+=1
			i+=1
			for j in range(j,startY,-1):
				nums[i][j]=count
				count+=1
			j-=1
			for i in range(i,startX,-1):
				nums[i][j]=count
				count+=1
			i-=1
			startX+=1
			startY+=1
			offset+=1
			loop-=1
		if n%2!=0:
			nums[n//2][n//2]=count
		return nums

	def printNums(self,nums:List[List[int]]):
		for num in nums:
			print(*num)

def main():
	solution:Solution=Solution()
	nums:List[List[int]]=solution.generateMatrix(5)
	solution.printNums(nums)

if __name__=="__main__":
	main()